File: sdk/go/zhtw/json_adapter.go
package zhtw

import (
	"fmt"
	"reflect"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// JSONAdapterError reports a stable failure code for safe JSON conversion.
type JSONAdapterError struct {
	Code    string
	Message string
}

func (err *JSONAdapterError) Error() string {
	return fmt.Sprintf("zhtw: JSON adapter %s: %s", err.Code, err.Message)
}

type jsonNode struct {
	kind      byte
	primitive string
	array     []jsonNode
	object    []jsonMember
}

type jsonMember struct {
	key   string
	value jsonNode
}

type jsonStringToken struct {
	start int
	end   int
	value string
}

type jsonScanner struct {
	text     string
	position int
	values   []jsonStringToken
}

func (scanner *jsonScanner) parse() (jsonNode, error) {
	scanner.skipWhitespace()
	node, err := scanner.parseValue(true)
	if err != nil {
		return jsonNode{}, err
	}
	scanner.skipWhitespace()
	if scanner.position != len(scanner.text) {
		return jsonNode{}, invalidJSON("trailing JSON content")
	}
	return node, nil
}

func (scanner *jsonScanner) parseValue(collectString bool) (jsonNode, error) {
	scanner.skipWhitespace()
	switch scanner.peekByte() {
	case '"':
		if _, err := scanner.parseString(collectString); err != nil {
			return jsonNode{}, err
		}
		return jsonNode{kind: 's'}, nil
	case '{':
		return scanner.parseObject()
	case '[':
		return scanner.parseArray()
	case 0:
		return jsonNode{}, invalidJSON("missing JSON value")
	default:
		start := scanner.position
		for current := scanner.peekByte(); current != 0 && !strings.ContainsRune(",]} \t\r\n", rune(current)); current = scanner.peekByte() {
			scanner.position++
		}
		primitive := scanner.text[start:scanner.position]
		if primitive != "true" && primitive != "false" && primitive != "null" && !validJSONNumber(primitive) {
			return jsonNode{}, invalidJSON("invalid JSON primitive")
		}
		return jsonNode{kind: 'p', primitive: primitive}, nil
	}
}

func (scanner *jsonScanner) parseObject() (jsonNode, error) {
	if err := scanner.expectByte('{'); err != nil {
		return jsonNode{}, err
	}
	scanner.skipWhitespace()
	members := make([]jsonMember, 0)
	keys := make(map[string]bool)
	if scanner.peekByte() == '}' {
		scanner.position++
		return jsonNode{kind: 'o', object: members}, nil
	}
	for {
		scanner.skipWhitespace()
		if scanner.peekByte() != '"' {
			return jsonNode{}, invalidJSON("JSON object key must be a string")
		}
		key, err := scanner.parseString(false)
		if err != nil {
			return jsonNode{}, err
		}
		if keys[key] {
			return jsonNode{}, jsonAdapterError("duplicate_key", "duplicate JSON object key")
		}
		keys[key] = true
		scanner.skipWhitespace()
		if err := scanner.expectByte(':'); err != nil {
			return jsonNode{}, err
		}
		value, err := scanner.parseValue(true)
		if err != nil {
			return jsonNode{}, err
		}
		members = append(members, jsonMember{key: key, value: value})
		scanner.skipWhitespace()
		if scanner.peekByte() == '}' {
			scanner.position++
			return jsonNode{kind: 'o', object: members}, nil
		}
		if err := scanner.expectByte(','); err != nil {
			return jsonNode{}, err
		}
	}
}

func (scanner *jsonScanner) parseArray() (jsonNode, error) {
	if err := scanner.expectByte('['); err != nil {
		return jsonNode{}, err
	}
	scanner.skipWhitespace()
	values := make([]jsonNode, 0)
	if scanner.peekByte() == ']' {
		scanner.position++
		return jsonNode{kind: 'a', array: values}, nil
	}
	for {
		value, err := scanner.parseValue(true)
		if err != nil {
			return jsonNode{}, err
		}
		values = append(values, value)
		scanner.skipWhitespace()
		if scanner.peekByte() == ']' {
			scanner.position++
			return jsonNode{kind: 'a', array: values}, nil
		}
		if err := scanner.expectByte(','); err != nil {
			return jsonNode{}, err
		}
	}
}

func (scanner *jsonScanner) parseString(collect bool) (string, error) {
	start := scanner.position
	if err := scanner.expectByte('"'); err != nil {
		return "", err
	}
	var decoded strings.Builder
	for scanner.position < len(scanner.text) {
		value, size := utf8.DecodeRuneInString(scanner.text[scanner.position:])
		if value == utf8.RuneError && size == 1 {
			return "", invalidJSON("invalid UTF-8 in JSON string")
		}
		scanner.position += size
		if value == '"' {
			result := decoded.String()
			if collect {
				scanner.values = append(scanner.values, jsonStringToken{start: start, end: scanner.position, value: result})
			}
			return result, nil
		}
		if value < 0x20 {
			return "", invalidJSON("unescaped JSON control character")
		}
		if value != '\\' {
			decoded.WriteRune(value)
			continue
		}
		escaped := scanner.nextByte()
		switch escaped {
		case '"', '\\', '/':
			decoded.WriteByte(escaped)
		case 'b':
			decoded.WriteByte('\b')
		case 'f':
			decoded.WriteByte('\f')
		case 'n':
			decoded.WriteByte('\n')
		case 'r':
			decoded.WriteByte('\r')
		case 't':
			decoded.WriteByte('\t')
		case 'u':
			first, err := scanner.parseHexQuad()
			if err != nil {
				return "", err
			}
			if first >= 0xd800 && first <= 0xdbff {
				if scanner.nextByte() != '\\' || scanner.nextByte() != 'u' {
					return "", invalidJSON("unpaired Unicode surrogate")
				}
				second, err := scanner.parseHexQuad()
				if err != nil || second < 0xdc00 || second > 0xdfff {
					return "", invalidJSON("unpaired Unicode surrogate")
				}
				decoded.WriteRune(utf16.DecodeRune(rune(first), rune(second)))
			} else if first >= 0xdc00 && first <= 0xdfff {
				return "", invalidJSON("unpaired Unicode surrogate")
			} else {
				decoded.WriteRune(rune(first))
			}
		default:
			return "", invalidJSON("invalid JSON escape")
		}
	}
	return "", invalidJSON("unterminated JSON string")
}

func (scanner *jsonScanner) parseHexQuad() (uint16, error) {
	var value uint16
	for i := 0; i < 4; i++ {
		current := scanner.nextByte()
		var digit byte
		switch {
		case current >= '0' && current <= '9':
			digit = current - '0'
		case current >= 'a' && current <= 'f':
			digit = current - 'a' + 10
		case current >= 'A' && current <= 'F':
			digit = current - 'A' + 10
		default:
			return 0, invalidJSON("invalid Unicode escape")
		}
		value = value*16 + uint16(digit)
	}
	return value, nil
}

func (scanner *jsonScanner) skipWhitespace() {
	for strings.ContainsRune(" \t\r\n", rune(scanner.peekByte())) {
		scanner.position++
	}
}

func (scanner *jsonScanner) peekByte() byte {
	if scanner.position >= len(scanner.text) {
		return 0
	}
	return scanner.text[scanner.position]
}

func (scanner *jsonScanner) nextByte() byte {
	value := scanner.peekByte()
	if value != 0 {
		scanner.position++
	}
	return value
}

func (scanner *jsonScanner) expectByte(expected byte) error {
	if scanner.nextByte() != expected {
		return invalidJSON("unexpected JSON token")
	}
	return nil
}

func (c *Converter) ConvertJSON(text string) (string, error) {
	originalScanner := jsonScanner{text: text}
	original, err := originalScanner.parse()
	if err != nil {
		return "", err
	}
	type replacement struct {
		start, end int
		value      string
	}
	replacements := make([]replacement, 0)
	for _, token := range originalScanner.values {
		target := c.Convert(token.value)
		if target != token.value {
			replacements = append(replacements, replacement{token.start, token.end, quoteJSONString(target)})
		}
	}
	if len(replacements) == 0 {
		return text, nil
	}
	var output strings.Builder
	output.Grow(len(text))
	lastEnd := 0
	for _, replacement := range replacements {
		output.WriteString(text[lastEnd:replacement.start])
		output.WriteString(replacement.value)
		lastEnd = replacement.end
	}
	output.WriteString(text[lastEnd:])
	convertedText := output.String()
	convertedScanner := jsonScanner{text: convertedText}
	converted, err := convertedScanner.parse()
	if err != nil {
		return "", err
	}
	if !reflect.DeepEqual(original, converted) {
		return "", jsonAdapterError("structure_changed", "converted JSON changed non-value structure")
	}
	return convertedText, nil
}

func quoteJSONString(value string) string {
	var output strings.Builder
	output.Grow(len(value) + 2)
	output.WriteByte('"')
	for _, current := range value {
		switch current {
		case '"':
			output.WriteString(`\"`)
		case '\\':
			output.WriteString(`\\`)
		case '\b':
			output.WriteString(`\b`)
		case '\f':
			output.WriteString(`\f`)
		case '\n':
			output.WriteString(`\n`)
		case '\r':
			output.WriteString(`\r`)
		case '\t':
			output.WriteString(`\t`)
		default:
			if current < 0x20 {
				fmt.Fprintf(&output, `\u%04x`, current)
			} else {
				output.WriteRune(current)
			}
		}
	}
	output.WriteByte('"')
	return output.String()
}

func validJSONNumber(value string) bool {
	position := 0
	if position < len(value) && value[position] == '-' {
		position++
	}
	if position >= len(value) {
		return false
	}
	if value[position] == '0' {
		position++
	} else if value[position] >= '1' && value[position] <= '9' {
		for position < len(value) && value[position] >= '0' && value[position] <= '9' {
			position++
		}
	} else {
		return false
	}
	if position < len(value) && value[position] == '.' {
		position++
		start := position
		for position < len(value) && value[position] >= '0' && value[position] <= '9' {
			position++
		}
		if position == start {
			return false
		}
	}
	if position < len(value) && (value[position] == 'e' || value[position] == 'E') {
		position++
		if position < len(value) && (value[position] == '+' || value[position] == '-') {
			position++
		}
		start := position
		for position < len(value) && value[position] >= '0' && value[position] <= '9' {
			position++
		}
		if position == start {
			return false
		}
	}
	return position == len(value)
}

func invalidJSON(message string) error {
	return jsonAdapterError("invalid_json", message)
}

func jsonAdapterError(code, message string) error {
	return &JSONAdapterError{Code: code, Message: message}
}

File: sdk/rust/zhtw/src/json_adapter.rs
use crate::error::{Error, Result};

#[derive(Debug, PartialEq, Eq)]
enum Node {
    StringValue,
    Primitive(String),
    Array(Vec<Node>),
    Object(Vec<(String, Node)>),
}

struct StringToken {
    start: usize,
    end: usize,
    value: String,
}

struct Scanner<'a> {
    text: &'a str,
    position: usize,
    values: Vec<StringToken>,
}

impl<'a> Scanner<'a> {
    fn new(text: &'a str) -> Self {
        Self {
            text,
            position: 0,
            values: Vec::new(),
        }
    }

    fn parse(&mut self) -> Result<Node> {
        self.skip_whitespace();
        let node = self.parse_value(true)?;
        self.skip_whitespace();
        if self.position != self.text.len() {
            return Err(invalid("trailing JSON content"));
        }
        Ok(node)
    }

    fn parse_value(&mut self, collect_string: bool) -> Result<Node> {
        self.skip_whitespace();
        match self.peek_byte() {
            Some(b'"') => {
                self.parse_string(collect_string)?;
                Ok(Node::StringValue)
            }
            Some(b'{') => self.parse_object(),
            Some(b'[') => self.parse_array(),
            Some(_) => {
                let start = self.position;
                while let Some(byte) = self.peek_byte() {
                    if b",]} \t\r\n".contains(&byte) {
                        break;
                    }
                    self.position += 1;
                }
                let primitive = &self.text[start..self.position];
                if !matches!(primitive, "true" | "false" | "null") && !valid_number(primitive) {
                    return Err(invalid("invalid JSON primitive"));
                }
                Ok(Node::Primitive(primitive.to_string()))
            }
            None => Err(invalid("missing JSON value")),
        }
    }

    fn parse_object(&mut self) -> Result<Node> {
        self.expect_byte(b'{')?;
        self.skip_whitespace();
        let mut members = Vec::new();
        let mut keys = std::collections::HashSet::new();
        if self.peek_byte() == Some(b'}') {
            self.position += 1;
            return Ok(Node::Object(members));
        }
        loop {
            self.skip_whitespace();
            if self.peek_byte() != Some(b'"') {
                return Err(invalid("JSON object key must be a string"));
            }
            let key = self.parse_string(false)?;
            if !keys.insert(key.clone()) {
                return Err(adapter_error("duplicate_key", "duplicate JSON object key"));
            }
            self.skip_whitespace();
            self.expect_byte(b':')?;
            members.push((key, self.parse_value(true)?));
            self.skip_whitespace();
            if self.peek_byte() == Some(b'}') {
                self.position += 1;
                return Ok(Node::Object(members));
            }
            self.expect_byte(b',')?;
        }
    }

    fn parse_array(&mut self) -> Result<Node> {
        self.expect_byte(b'[')?;
        self.skip_whitespace();
        let mut values = Vec::new();
        if self.peek_byte() == Some(b']') {
            self.position += 1;
            return Ok(Node::Array(values));
        }
        loop {
            values.push(self.parse_value(true)?);
            self.skip_whitespace();
            if self.peek_byte() == Some(b']') {
                self.position += 1;
                return Ok(Node::Array(values));
            }
            self.expect_byte(b',')?;
        }
    }

    fn parse_string(&mut self, collect: bool) -> Result<String> {
        let start = self.position;
        self.expect_byte(b'"')?;
        let mut decoded = String::new();
        while self.position < self.text.len() {
            let ch = self.text[self.position..].chars().next().unwrap();
            self.position += ch.len_utf8();
            if ch == '"' {
                if collect {
                    self.values.push(StringToken {
                        start,
                        end: self.position,
                        value: decoded.clone(),
                    });
                }
                return Ok(decoded);
            }
            if ch <= '\u{1f}' {
                return Err(invalid("unescaped JSON control character"));
            }
            if ch != '\\' {
                decoded.push(ch);
                continue;
            }
            let escaped = self
                .next_byte()
                .ok_or_else(|| invalid("unterminated JSON escape"))?;
            match escaped {
                b'"' => decoded.push('"'),
                b'\\' => decoded.push('\\'),
                b'/' => decoded.push('/'),
                b'b' => decoded.push('\u{8}'),
                b'f' => decoded.push('\u{c}'),
                b'n' => decoded.push('\n'),
                b'r' => decoded.push('\r'),
                b't' => decoded.push('\t'),
                b'u' => {
                    let first = self.parse_hex_quad()?;
                    let codepoint = if (0xd800..=0xdbff).contains(&first) {
                        if self.next_byte() != Some(b'\\') || self.next_byte() != Some(b'u') {
                            return Err(invalid("unpaired Unicode surrogate"));
                        }
                        let second = self.parse_hex_quad()?;
                        if !(0xdc00..=0xdfff).contains(&second) {
                            return Err(invalid("unpaired Unicode surrogate"));
                        }
                        0x10000 + (((first - 0xd800) as u32) << 10) + (second - 0xdc00) as u32
                    } else if (0xdc00..=0xdfff).contains(&first) {
                        return Err(invalid("unpaired Unicode surrogate"));
                    } else {
                        first as u32
                    };
                    decoded
                        .push(char::from_u32(codepoint).ok_or_else(|| invalid("invalid Unicode"))?);
                }
                _ => return Err(invalid("invalid JSON escape")),
            }
        }
        Err(invalid("unterminated JSON string"))
    }

    fn parse_hex_quad(&mut self) -> Result<u16> {
        let mut value = 0u16;
        for _ in 0..4 {
            let byte = self
                .next_byte()
                .ok_or_else(|| invalid("short Unicode escape"))?;
            let digit = (byte as char)
                .to_digit(16)
                .ok_or_else(|| invalid("invalid Unicode escape"))?;
            value = value * 16 + digit as u16;
        }
        Ok(value)
    }

    fn skip_whitespace(&mut self) {
        while self
            .peek_byte()
            .is_some_and(|byte| b" \t\r\n".contains(&byte))
        {
            self.position += 1;
        }
    }

    fn peek_byte(&self) -> Option<u8> {
        self.text.as_bytes().get(self.position).copied()
    }

    fn next_byte(&mut self) -> Option<u8> {
        let value = self.peek_byte()?;
        self.position += 1;
        Some(value)
    }

    fn expect_byte(&mut self, expected: u8) -> Result<()> {
        if self.next_byte() != Some(expected) {
            return Err(invalid("unexpected JSON token"));
        }
        Ok(())
    }
}

pub(crate) fn convert_json_values(
    text: &str,
    converter: impl Fn(&str) -> String,
) -> Result<String> {
    let mut original_scanner = Scanner::new(text);
    let original = original_scanner.parse()?;
    let mut replacements = Vec::new();
    for token in original_scanner.values {
        let target = converter(&token.value);
        if target != token.value {
            replacements.push((token.start, token.end, quote(&target)));
        }
    }
    if replacements.is_empty() {
        return Ok(text.to_string());
    }
    let mut output = String::with_capacity(text.len());
    let mut last_end = 0;
    for (start, end, replacement) in replacements {
        output.push_str(&text[last_end..start]);
        output.push_str(&replacement);
        last_end = end;
    }
    output.push_str(&text[last_end..]);
    let converted = Scanner::new(&output).parse()?;
    if converted != original {
        return Err(adapter_error(
            "structure_changed",
            "converted JSON changed non-value structure",
        ));
    }
    Ok(output)
}

pub(crate) fn quote(value: &str) -> String {
    let mut output = String::with_capacity(value.len() + 2);
    output.push('"');
    for ch in value.chars() {
        match ch {
            '"' => output.push_str("\\\""),
            '\\' => output.push_str("\\\\"),
            '\u{8}' => output.push_str("\\b"),
            '\u{c}' => output.push_str("\\f"),
            '\n' => output.push_str("\\n"),
            '\r' => output.push_str("\\r"),
            '\t' => output.push_str("\\t"),
            ch if ch <= '\u{1f}' => output.push_str(&format!("\\u{:04x}", ch as u32)),
            ch => output.push(ch),
        }
    }
    output.push('"');
    output
}

fn valid_number(value: &str) -> bool {
    let bytes = value.as_bytes();
    let mut i = 0;
    if bytes.get(i) == Some(&b'-') {
        i += 1;
    }
    match bytes.get(i) {
        Some(b'0') => i += 1,
        Some(b'1'..=b'9') => {
            i += 1;
            while bytes.get(i).is_some_and(u8::is_ascii_digit) {
                i += 1;
            }
        }
        _ => return false,
    }
    if bytes.get(i) == Some(&b'.') {
        i += 1;
        let start = i;
        while bytes.get(i).is_some_and(u8::is_ascii_digit) {
            i += 1;
        }
        if i == start {
            return false;
        }
    }
    if matches!(bytes.get(i), Some(b'e' | b'E')) {
        i += 1;
        if matches!(bytes.get(i), Some(b'+' | b'-')) {
            i += 1;
        }
        let start = i;
        while bytes.get(i).is_some_and(u8::is_ascii_digit) {
            i += 1;
        }
        if i == start {
            return false;
        }
    }
    i == bytes.len()
}

fn invalid(message: &str) -> Error {
    adapter_error("invalid_json", message)
}

fn adapter_error(code: &'static str, message: &str) -> Error {
    Error::JsonAdapter {
        code,
        message: message.to_string(),
    }
}
